### PR TITLE
Fix parsing extra heartbeats

### LINKED ISSUE #583: Parsing extra param int as str always fails
When parsing extra params json fails because an int was jsonified as str, a second attempt is made with a different struct accepting str. For example, when `cursorpos` is a string instead of int:

`{"timestamp": 123.456, "cursorpos": "42"}`

But then parsing timestamp fails because it's not rendered as string:

```
"failed to read extra heartbeats: failed to json decode: failed to json decode from data \"[{\\\"cursorpos\\\": \\\"42\\\", \\\"timestamp\\\": 1637185415.093967}]\\n\": json: cannot unmarshal string into Go struct field .cursorpos of type int: failed to json decode again, accepting string values: failed to json decode from data \"[{\\\"cursorpos\\\": \\\"42\\\", \\\"timestamp\\\": 1637185415.093967}]\\n\": json: cannot unmarshal number into Go struct field .timestamp of type string"
```

This all-or-nothing approach is too brittle. Instead, we must be able to parse any int or float param as either string or number without errors.

### PR DESCRIPTION
**Pull request by gandarez**
This PR fixes extra heartbeats parsing to accept any kind of data. Strong types were replaced by `interface{}` in order to be more flexible in terms of received json.

fixes https://github.com/wakatime/wakatime-cli/issues/583

### CODE REVIEW

**Comment by dron22** [comment]
Hey @gandarez! I will review this later today at about 8pm CET :+1:

File: cmd/legacy/heartbeat/params.go [hunk 1]
```diff
@@ -305,19 +295,19 @@ func readExtraHeartbeats() ([]heartbeat.Heartbeat, error) {
 func parseExtraHeartbeat(data string) ([]heartbeat.Heartbeat, error) {
 	var incoming []struct {
 		Category          heartbeat.Category `json:"category"`
-		CursorPosition    *int               `json:"cursorpos"`
+		CursorPosition    interface{}        `json:"cursorpos"`
 		Entity            string             `json:"entity"`
 		EntityType        string             `json:"entity_type"`
 		Type              string             `json:"type"`
-		IsWrite           *bool              `json:"is_write"`
```
Comment (alanhamlett) [inline cmd/legacy/heartbeat/params.go:312]: What values will be accepted as true/false?

Comment (gandarez) [reply cmd/legacy/heartbeat/params.go:312]: Any bool or string. `true`, `false`, `"true"`and `"false"`.

File: cmd/legacy/heartbeat/params.go [hunk 2]
```diff
@@ -344,128 +334,97 @@ func parseExtraHeartbeat(data string) ([]heartbeat.Heartbeat, error) {
...
-
-	var heartbeats []heartbeat.Heartbeat
-
-	for _, h := range incoming {
 		var cursorPosition *int
 
-		if h.CursorPosition != nil {
-			parsed, err := strconv.Atoi(*h.CursorPosition)
+		switch cursorPositionVal := h.CursorPosition.(type) {
+		case float64:
```
Comment (dron22) [inline cmd/legacy/heartbeat/params.go:340]: I guess we should add a case for `int` type as well. This way, we support `float64`, `int` and `string`. This applies for `lineno` and `lines` values as well.

File: cmd/legacy/heartbeat/params.go [hunk 3]
```diff
@@ -344,128 +334,97 @@ func parseExtraHeartbeat(data string) ([]heartbeat.Heartbeat, error) {
...
 		}
 
-		var timestamp float64
+		var time float64
 
-		switch {
-		case h.Time != nil:
-			parsed, err := strconv.ParseFloat(*h.Time, 64)
+		switch timeVal := h.Time.(type) {
+		case float64:
```
Comment (dron22) [inline cmd/legacy/heartbeat/params.go:396]: For `time` property, support for case `int` type would make it more robust as well.

File: cmd/legacy/heartbeat/params.go [hunk 4]
```diff
@@ -344,128 +334,97 @@ func parseExtraHeartbeat(data string) ([]heartbeat.Heartbeat, error) {
...
-			timestamp = parsed
-		case h.Timestamp != nil:
-			parsed, err := strconv.ParseFloat(*h.Timestamp, 64)
+			time = val
+		}
+
+		var timestamp float64
+
+		switch timestampVal := h.Timestamp.(type) {
+		case float64:
```
Comment (dron22) [inline cmd/legacy/heartbeat/params.go:410]: For `timestamp` property, support for case `int` type would make it more robust as well.

**Review by dron22 — DISMISSED** [review]
In general LGTM, nice usage of type switches :+1: To make it completely robust, I think we need additional cases for `int` types.

In general adding test cases to cover this new behavior would be good.

**Comment by gandarez** [comment]
I think we've covered all use cases for extra heartbeats which are string or the type itself (`TestLoadParams_ExtraHeartbeats` and `TestLoadParams_ExtraHeartbeats_WithStringValues`). When Go unmarshals for `interface{}` then it's converted to the following types under the hood. It's documented [here](https://pkg.go.dev/encoding/json#Unmarshal). So we actually don't need to add an extra case to verify `int` type when parsing for numbers because it'll always be `float64`.

```
bool, for JSON booleans
float64, for JSON numbers
string, for JSON strings
[]interface{}, for JSON arrays
map[string]interface{}, for JSON objects
nil for JSON null
```

**Comment by alanhamlett** [comment]
> `float64, for JSON numbers`

So it always uses `float64` when parsing JSON int and floats?

**Comment by alanhamlett** [comment]
Let's add tests for all possible inputs and then LGTM!

**Comment by gandarez** [comment]
All possible inputs are already covered by `TestLoadParams_ExtraHeartbeats` and `TestLoadParams_ExtraHeartbeats_WithStringValues`. Do you guys see other ways to tests?

**Comment by alanhamlett** [comment]
> All possible inputs are already covered by `TestLoadParams_ExtraHeartbeats` and `TestLoadParams_ExtraHeartbeats_WithStringValues`. Do you guys see other ways to tests?

We could change `"lineno": "42"` to `"lineno": 42` but only in one of the two heartbeats here:
https://github.com/wakatime/wakatime-cli/blob/970e80e2139dcff9e772bf79eab4fb9046f6f790/cmd/legacy/heartbeat/testdata/extra_heartbeats_with_string_values.json#L1

That would test for the actual bug in #583 which is triggered from mixing string and integer values in a heartbeat.

**Comment by gandarez** [comment]
@alanhamlett because we changed from strong types to interfaces and we do have tests for both I don't see any advantage of changing one of them from string to the real type. Those scenarios are already covered because in the past we had two independent structs, where one for the strong types and only accepted real values and another one to accept only strings. The new scenario fixes this approach and accepts any kind of value. We need to make sure we're testing both cases for each (interfaced) field.

**Review by alanhamlett — APPROVED** [review]

**Comment by alanhamlett** [comment]
Yea, that makes sense. In that case, looks great!

**Pull request merged**